### PR TITLE
Fix wave restart for cold start and add ic version file

### LINKED ISSUE #3109: WW3 ICs are not read for HR like experiements.
### What is wrong?

The gfs_stage_ic job copies wave ICs: 

` ^[[38;21m2024-11-17 13:45:38,610 - INFO     - file_utils  : Copied /lfs/h2/emc/couple/noscrub/jessica.meixner/WaveUglo15km/ICDIR/Opt3/gfs.20200912/18/model/wave/restart/20200913.000000.restart.ww3 to /lfs/h2/emc/couple/noscrub/jessica.meixner/WaveUglo15km/Opt3Sept/COMROOT/Opt3Sept/gfs.20200912/18//model/wave/restart^[[0m`

But then the forecast job does not find ICs: 
```
+ exglobal_forecast.sh[124]: WW3_postdet
+ forecast_postdet.sh[327]: echo 'SUB WW3_postdet: Linking input data for WW3'
SUB WW3_postdet: Linking input data for WW3
+ forecast_postdet.sh[329]: local ww3_grid first_ww3_restart_out ww3_restart_file
+ forecast_postdet.sh[331]: [[ .false. == \.\t\r\u\e\. ]]
+ forecast_postdet.sh[357]: echo 'WW3 will start from rest!'
WW3 will start from rest!
+ forecast_postdet.sh[358]: first_ww3_restart_out=2020091300
+ forecast_postdet.sh[362]: local ww3_restart_file
+ forecast_postdet.sh[364]: (( vdate = first_ww3_restart_out ))
+ forecast_postdet.sh[364]: (( vdate <= forecast_end_cycle ))
+ forecast_postdet.sh[366]: ww3_restart_file=20200913.000000.restart.ww3
+ forecast_postdet.sh[367]: /bin/ln -sf /lfs/h2/emc/stmp/jessica.meixner/RUNDIRS/Opt3Sept/gfs.2020091300/gfsfcst.2020091300/restart/WW3_RESTART/20200913.000000.restart.ww3 20200913.000000.restart.ww3
```

### What should have happened?

ICs should be used in the HR prototype-like cases.   

### What machines are impacted?

All or N/A

### What global-workflow hash are you using?

Techncially it's a branch, but this also appears to have happend in HR4 tag 

### Steps to reproduce

I'm running my branch here: https://github.com/JessicaMeixner-NOAA/global-workflow/tree/feature/uglo_15km 
and staging my own IC directory. 

So from: /lfs/h2/emc/couple/noscrub/jessica.meixner/WaveUglo15km/global-workflow/workflow 
using the script coupled.sh to set up experiments. 

### Additional information

I checked HR4 output and there are no wave ICs as HS is 0 at the first output (haven't confirmed from log though). 

Have pinged @sbanihash to be sure to check GEFS recent experiment results.  

Have not yet checked low-res Ci tests, but there's likely an example use-case that could be used for debugging. 

### Do you have a proposed solution?

Trying to figure out logic https://github.com/NOAA-EMC/global-workflow/blob/develop/ush/forecast_postdet.sh#L330-L359 and in the stage_ic job to figure out where the disconnect is.

### PR DESCRIPTION
**Pull request by WalterKolczynski-NOAA**
# Description
The stage job was incorrectly putting wave restarts into the gfs directory. The forecast job looks for them in the gdas directory, so this is updated.

Additionally, the restarts were also not being copied from the staged directory to `$DATA`, so now they are. The process is identical to that of non-RERUN warm starts, so the code is refactored a bit to avoid duplication.

As part of updating the ICs with the new restart location, an IC version file is added to support different version numbers for different IC directories. Unlike other version files, this one uses an associative array rather than individual variables.

Resolves #3109

# Type of change
- [x] Bug fix (fixes something broken)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO
- Does this change require an update to any of the following submodules? NO

# How has this been tested?
- C48_S2SW test on Hercules
- C48_S2SWA_gefs test on Hercules

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have documented my code, including function, input, and output descriptions
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] This change is covered by an existing CI test or a new one has been added
- [x] Any new scripts have been added to the .github/CODEOWNERS file with owners
- [x] I have made corresponding changes to the system documentation if necessary

### CODE REVIEW

**Comment by WalterKolczynski-NOAA** [comment]
This won't work yet because I need to move all the wave restarts in ICSDIR.

**Comment by JessicaMeixner-NOAA** [comment]
> This won't work yet because I need to move all the wave restarts in ICSDIR.

I have a directory of just 1 IC test so I can get it moved and tested from there pretty easily.  Quick question, I should be staging in the previous cycle gdas, correct?

**Comment by WalterKolczynski-NOAA** [comment]
> > This won't work yet because I need to move all the wave restarts in ICSDIR.
> 
> I have a directory of just 1 IC test so I can get it moved and tested from there pretty easily. Quick question, I should be staging in the previous cycle gdas, correct?

Yes. Should be the same as we have now, except gfs ➡ gdas

**Comment by JessicaMeixner-NOAA** [comment]
@WalterKolczynski-NOAA - I have  test running.  It's not "clean" in the sense that I merged your changes into my branch - but the stage-IC job succeeded and forecast is in the queue.  I'll report in the morning and thank you so much for a quick fix to this problem!

File: parm/stage/master_gfs.yaml.j2
Comment (KateFriedman-NOAA) [inline parm/stage/master_gfs.yaml.j2:111]: If `previous_cycle_and_run_dict` is not going to be used here then lines 97-101 should also be removed. Only the wave IC staging was using it based on the pre-refactored staging job behavior and the ICs on disk.

This seems to be another cold-start quirk, which wasn't captured in the CI tests at the time of refactoring. Thanks for resolving.

Comment (JessicaMeixner-NOAA) [reply parm/stage/master_gfs.yaml.j2:111]: @KateFriedman-NOAA  these ICs were used in CI tests at one point. I'm not sure where the bug(s) crept in.  C48_S2SW should be using and IC for the wave component for example.

Comment (WalterKolczynski-NOAA) [reply parm/stage/master_gfs.yaml.j2:111]: Removed

**Comment by JessicaMeixner-NOAA** [comment]
@WalterKolczynski-NOAA - My test from last night has a wave IC!!!!  Thank you again for this quick update.  Even just having this branch means we can move forward with getting some runs comparing different physics options for the wave model now.

**Comment by WalterKolczynski-NOAA** [comment]
New IC versions have been created for the relocated wave restarts. This required adding an IC versions file.

Note: the high-res cases (C768/C1152) likely will still not work. They use the wave grid name as the restart suffix, but for single-grid waves, ending in `.ww3` is expected. Unlike the other directories, where I made this change previously by adding symlinks, I can't make create links for the high-res wave restarts because they point to the prototype ICs owned by the climate group. (I guess I could redo it to just point at the file instead of the directory, but I haven't.)

**Comment by JessicaMeixner-NOAA** [comment]
> New IC versions have been created for the relocated wave restarts. This required adding an IC versions file.
> 
> Note: the high-res cases (C768/C1152) likely will still not work. They use the wave grid name as the restart suffix, but for single-grid waves, ending in `.ww3` is expected. Unlike the other directories, where I made this change previously by adding symlinks, I can't make create links for the high-res wave restarts because they point to the prototype ICs owned by the climate group. (I guess I could redo it to just point at the file instead of the directory, but I haven't.)

I have access to add additional links to files owned by climate group or can coordinate this being found. Can you let me know which directories?

**Review by KateFriedman-NOAA — DISMISSED** [review]
Looks good, thanks for the IC updates/fixes @WalterKolczynski-NOAA ! Approve pending successful completion of CI testing.

File: versions/ic.ver
Comment (KateFriedman-NOAA) [inline versions/ic.ver:8]: Oh, the `icsdir` paths in the CI test yamls ([example](https://github.com/NOAA-EMC/global-workflow/blob/develop/ci/cases/pr/C48mx500_3DVarAOWCDA.yaml#L12)) will also need to be updated for the resolutions that have a new timestamp.

Comment (JessicaMeixner-NOAA) [reply versions/ic.ver:8]: We might want to update the title of this PR to reflect the expanded scope of the IC dirs that this touches to get more eyes on things that are changing. This PR is going beyond just wave ICs at this point - although the main reason for the PR, and the fix is greatly appreciated.

**Review by JessicaMeixner-NOAA — COMMENTED** [review]
I've tested the wave parts of these changes and approve those changes.  This PR however touches a lot more than just wave ICs with the IC version change, so maybe posting the output of the CI for more people to look at before giving the official approval might be a good idea?

**Comment by JessicaMeixner-NOAA** [comment]
@WalterKolczynski-NOAA  - Can you remind me again of which machine which ICs I need to update for the wave model? 


I wanted to say you said an HR3?  

I can also work with @jiandewang to get this updated on WCOSS2 and run an HR4 test there as part of the review as well.

**Comment by JessicaMeixner-NOAA** [comment]
> @WalterKolczynski-NOAA - Can you remind me again of which machine which ICs I need to update for the wave model?
> 
> I wanted to say you said an HR3?
> 
> I can also work with @jiandewang to get this updated on WCOSS2 and run an HR4 test there as part of the review as well.

@WalterKolczynski-NOAA i just looked on wcoss2, this looks update?  So can I run an HR4 like test there now?  In the meantime, I'm not sure what you need changed otherwise so I'll wait for your response on testing and what else needs changed.

**Comment by WalterKolczynski-NOAA** [comment]
> @WalterKolczynski-NOAA - Can you remind me again of which machine which ICs I need to update for the wave model?
> 
> I wanted to say you said an HR3?
> 
> I can also work with @jiandewang to get this updated on WCOSS2 and run an HR4 test there as part of the review as well.

@JessicaMeixner-NOAA
HR3, on all machines. Just need to add a symlink in each of the wave restart directories that uses the `.ww3` suffix instead of `.uglo_m1g16`. This code run should do the trick:
```
20:35:04 HR3marine/> \
for restart in $(find /scratch1/NCEPDEV/climate/role.ufscpara/IC/HR3marine -name '*restart.uglo_m1g16'); do
  dir=$(dirname ${restart})
  target=$(basename ${restart})
  link=${target/%.uglo_m1g16/.ww3}
  cd ${dir}
  ln -s ${target} ${link}
done
```

Needs to be repeated on all machines.

(These changes are related to a previous PR fixing wave restarts #3009.)

**Comment by JessicaMeixner-NOAA** [comment]
I've made the updates on hera and orion/hercules.  


@jiandewang will have to make them on WCOSS2: 

```
for restart in $(find  /lfs/h2/emc/couple/noscrub/Jiande.Wang/IC/HR3marine -name '*restart.uglo_m1g16'); do
  dir=$(dirname ${restart})
  target=$(basename ${restart})
  link=${target/%.uglo_m1g16/.ww3}
  cd ${dir}
  ln -s ${target} ${link}
done
```

**Comment by jiandewang** [comment]
I will do on wcoss2 shortly (but in meeting now)

**Comment by jiandewang** [comment]
@JessicaMeixner-NOAA done on wcoss2 dev and prod machines.
@WalterKolczynski-NOAA your script works perfectly

File: versions/ic.ver
Comment (WalterKolczynski-NOAA) [reply versions/ic.ver:8]: Paths updated, though we probably don't need to specify the icsdir at all in these.

**Comment by WalterKolczynski-NOAA** [comment]
Awaiting approval before running CI on the other machines.

**Comment by JessicaMeixner-NOAA** [comment]
@WalterKolczynski-NOAA  - Can you provide an overview of the wave ICs that are used in CI tests and what changes were made in staged ICs for those ICs?   

The changes we made for HR3marine is fine because there was just one wave grid there, but for other tests it's unclear if we had multiple grids - and it seems like changes to wave grids were made that are now breaking other CI tests (see: https://github.com/NOAA-EMC/global-workflow/pull/3115#issuecomment-2494034414).

**Comment by JessicaMeixner-NOAA** [comment]
@WalterKolczynski-NOAA  My clean test run of a C1152 case ran successfully on wcoss2 and included a wave IC.  However, I have some concerns about how ICs might have been staged as it seems there are failures potentially related to these changes.

**Comment by DavidHuber-NOAA** [comment]
@JessicaMeixner-NOAA @WalterKolczynski-NOAA  It looks like the wave restarts are not staged properly on Hera, at least for the C48C48mx500 resolution.  The file `20210323.120000.restart.ww3` is missing from `/scratch1/NCEPDEV/global/glopara/data/ICSDIR/C48C48mx500/20240610/gefs.20210323/06/mem00
0/model/wave/restart` but present on Hercules under `/work/noaa/global/glopara/data/ICSDIR/C48C48mx500/20240610/gefs.20210323/06/mem000/model/wave/restart`.

**Comment by JessicaMeixner-NOAA** [comment]
Okay here are a list of CI tests that should be using WW3 ICs: 

https://github.com/NOAA-EMC/global-workflow/blob/develop/ci/cases/weekly/C384_S2SWA.yaml - gfs C384 2016070100 
https://github.com/NOAA-EMC/global-workflow/blob/develop/ci/cases/hires/C1152_S2SW.yaml gfs C1152 2019120300
https://github.com/NOAA-EMC/global-workflow/blob/develop/ci/cases/hires/C768_S2SW.yaml gfs C768 2019120300 
https://github.com/NOAA-EMC/global-workflow/blob/develop/ci/cases/pr/C48_S2SW.yaml gfs C48 2021032312
https://github.com/NOAA-EMC/global-workflow/blob/develop/ci/cases/pr/C48_S2SWA_gefs.yaml gefs C48 2021032312 
https://github.com/NOAA-EMC/global-workflow/blob/develop/ci/cases/pr/C96_S2SWA_gefs_replay_ics.yaml gefs C96 2020110100  {{ 'ICSDIR_ROOT' | getenv }}/C96mx100/20240610 

We can then use this list to back-trace, what grid and what date, and where an IC for waves should be pre and post this update and make sure things are as expected. 


And agreed @DavidHuber-NOAA - there do seem to be issues.

**Comment by JessicaMeixner-NOAA** [comment]
I'm concerned we might have a conflict for ICs for 
https://github.com/NOAA-EMC/global-workflow/blob/develop/ci/cases/pr/C48_S2SW.yaml gfs C48 2021032312
and
https://github.com/NOAA-EMC/global-workflow/blob/develop/ci/cases/pr/C48_S2SWA_gefs.yaml gefs C48 2021032312

as of right now I think these would be using the same IC directory, but we have 2 different wave grids, still working on details - but want to report the potential issue now.

**Comment by WalterKolczynski-NOAA** [comment]
> @WalterKolczynski-NOAA - Can you provide an overview of the wave ICs that are used in CI tests and what changes were made in staged ICs for those ICs?
> 
> The changes we made for HR3marine is fine because there was just one wave grid there, but for other tests it's unclear if we had multiple grids - and it seems like changes to wave grids were made that are now breaking other CI tests (see: [#3115 (comment)](https://github.com/NOAA-EMC/global-workflow/pull/3115#issuecomment-2494034414)).

A few symlinks were inadvertently broken in places I didn't intend to modify. I've restored them.

**Comment by WalterKolczynski-NOAA** [comment]
> I'm concerned we might have a conflict for ICs for https://github.com/NOAA-EMC/global-workflow/blob/develop/ci/cases/pr/C48_S2SW.yaml gfs C48 2021032312 and https://github.com/NOAA-EMC/global-workflow/blob/develop/ci/cases/pr/C48_S2SWA_gefs.yaml gefs C48 2021032312
> 
> as of right now I think these would be using the same IC directory, but we have 2 different wave grids, still working on details - but want to report the potential issue now.

GFS and GEFS have different directories within that version because `${RUN}.${PDY}` is part of the path.

**Comment by WalterKolczynski-NOAA** [comment]
```
[role.glopara@hfe05 ICSDIR]$ tree C48C48mx500/20241120 -L 2 
C48C48mx500/20241120
|-- enkfgdas.20210323
|   |-- 06
|   `-- 12
|-- enkfgdas.20210324
|   |-- 06
|   `-- 12
|-- gdas.20210323
|   |-- 06
|   `-- 12
|-- gdas.20210324
|   |-- 06
|   `-- 12
|-- gefs.20210323
|   |-- 06
|   `-- 12
`-- gfs.20210323
    `-- 12
```

**Comment by JessicaMeixner-NOAA** [comment]
Thanks for sharing that @WalterKolczynski-NOAA  .  I'm glad that the old tests should still work.

**Comment by JessicaMeixner-NOAA** [comment]
@WalterKolczynski-NOAA where would I find the path of output for regression tests?

**Comment by JessicaMeixner-NOAA** [comment]
@WalterKolczynski-NOAA:
-bash: cd: /work2/noaa/global/CI/HERCULES/3112: Permission denied

I've tested these changes on my end.  I'm okay w/these changes on the basis that we run CI tests.  I'm slightly concerned that we aren't confirming that non-wave-gfs tests aren't replicating answers though.  I can try to at least look at the log files to do a bit more checking since my review was requested again.

**Comment by WalterKolczynski-NOAA** [comment]
> @WalterKolczynski-NOAA: -bash: cd: /work2/noaa/global/CI/HERCULES/3112: Permission denied
> 
> I've tested these changes on my end. I'm okay w/these changes on the basis that we run CI tests. I'm slightly concerned that we aren't confirming that non-wave-gfs tests aren't replicating answers though. I can try to at least look at the log files to do a bit more checking since my review was requested again.

Your review was requested again because you haven't actually approved it. But this is close enough to start CI.

**Comment by WalterKolczynski-NOAA** [comment]
CI Tests set up to run in /lfs/h2/emc/ptmp/walter.kolczynski/PR/PR_3112/RUNTESTS on WCOSS

**Review by KateFriedman-NOAA — APPROVED** [review]
Thanks @WalterKolczynski-NOAA !

**Review by DavidHuber-NOAA — APPROVED** [review]

**Comment by danholdaway** [comment]
@WalterKolczynski-NOAA @JessicaMeixner-NOAA might this change slightly alter the forecast fields when cycling?

**Comment by JessicaMeixner-NOAA** [comment]
> @WalterKolczynski-NOAA @JessicaMeixner-NOAA might this change slightly alter the forecast fields when cycling?

I have not cycled w/waves yet  --- i'm not sure if we have all the proper things in place or not. I am over-due to look at this.  Are you cycling w/waves and seeing issues?

**Comment by JessicaMeixner-NOAA** [comment]
@danholdaway - To my knowledge, this PR should only change answers with tests that have waves included as it would have included a wave IC and therefor change answers.  To my knowledge, there are no CI tests that cycle with waves, so these tests should not have changed answers from this PR based on my understanding.  @WalterKolczynski-NOAA  would know more.

**Comment by danholdaway** [comment]
Thanks @JessicaMeixner-NOAA, we should be OK with what we're testing actually. Looking back through PRs to see if there's anything obvious that might have changed forecast output.

**Pull request merged**
